source/versioned-api/versioned-api.rst

```diff
@@ -135,12 +135,15 @@ rationale for more details.
 Sending Declared API Version to the Server
 ------------------------------------------
 
-The declared API version MUST be sent to the server with every command, with the
-exception of the cases listed below. Drivers MUST NOT use a server's reported
-``maxWireVersion`` to decide whether it supports the versioned API. The server
-will reply with an error if the declared API version is not supported. If the
-user does not declare an API version, the driver MUST NOT send any API
-versioning options to the server.
+The declared API version MUST be sent to any server that supports ``OP_MSG``
+as part of the command request, with the exception of the cases listed below.
+Drivers MUST NOT use a server's reported ``maxWireVersion`` to decide whether
+it supports the versioned API. The server will reply with an error if the
+declared API version is not supported. If the user does not declare an API
+version, the driver MUST NOT send any API versioning options to the server.
+Note that per the ``OP_MSG`` spec the driver is required to use ``OP_MSG``
+when communicating with an ``OP_MSG`` capable server. Versioned API options
+MUST NOT be sent to pre-``OP_MSG`` servers.
```

Comment:
> Versioned API options MUST NOT be sent to pre-``OP_MSG`` servers.

So does this mean that the ServerApi option is ignored when talking to <=3.4 servers?

Reply:
I can see how this is still potentially confusing. The intention of the PR was to not require any changes when OP_MSG is not used. Are you sending server API parameters using legacy protocol? If yes how do you send them in e.g. OP_QUERY?

Reply:
> Are you sending server API parameters using legacy protocol?

Yes. The OP_QUERY format supports sending arbitrary command arguments in the same way that OP_MSG does (the runCommand helper works with OP_QUERY after all).  But now that you mention it, non-command based protocols do not support these arguments at all, like OP_INSERT or a find() on a <=3.0 server using OP_QUERY (it's a bit confusing to talk because OP_QUERY has 2 uses). Now I understand what this is trying to clarify.

I am under the impression that we want every operation on <=4.4 to raise an error when versioned api configured. Personally I think we should do DRIVERS-1484 and ban sending apiVersion to <=4.4 servers altogether. Then all non-OP_MSG would be banned (except of course the isMaster handshake command which still uses OP_QUERY).

Reply:
So, when pymongo is connected to a 2.6 server, does it send api version with every command as the spec requires?

Reply:
Every command, yes. But pymongo won't send it with non-command messages like OP_INSERT or find on <=3.0.

Reply:
Okay, I see. I haven't been around when the legacy op codes were used so I was asking around how they are supposed to be handled and as far as I know there wasn't a helpful list of which op codes should have api parameters added to them and which shouldn't.

If the idea here is to have every driver figure out how the legacy op codes work independently I'll work on that, otherwise the thought with this PR was to exempt pre-3.6 servers (and thus legacy op codes) from needing to be handled by the drivers.